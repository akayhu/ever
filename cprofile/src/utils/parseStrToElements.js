import compose from './compose';
import n12br from './nl2br';
import ReactHtmlParser from 'components/react-html-parser';

// HTML String 轉成 React Elements
const parseStrToElements = str => {
	if (!str || typeof str !== 'string') return null;

	return compose(
		ReactHtmlParser,
		n12br
	)(str);
};

export default parseStrToElements;
