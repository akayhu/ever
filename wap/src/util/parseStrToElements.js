import compose from 'src/util/compose';
import n12br from 'src/util/n12br';
// import DOMPurify from 'dompurify';
import ReactHtmlParser from 'react-html-parser';
// import { canUseDOM } from 'exenv';

// HTML String 轉成 React Elements
const parseStrToElements = (str) => {
	if (!str || typeof str !== 'string') return null;

	// 因為 DOMPurify 會移除非正規的自訂 attribute，考量到要盤點眾多可能性（尤其是轉貼連結的顯示），決定暫且移除
	/* if (!canUseDOM) {
		const { JSDOM } = require('jsdom');
		const window = (new JSDOM('')).window;
		const ServerSideDOMPurify = DOMPurify(window);
		return compose(ReactHtmlParser, n12br, ServerSideDOMPurify.sanitize)(str);
	} */

	return compose(ReactHtmlParser, n12br/* , DOMPurify.sanitize */)(str);
};

export default parseStrToElements;
