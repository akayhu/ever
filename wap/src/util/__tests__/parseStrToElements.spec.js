/* eslint-disable */
import parseStrToElements from '../parseStrToElements';

describe('parseStrToElements', () => {
    it('非字串的輸入, expect null', () => {
        expect(parseStrToElements(['a'])).toBe(null);
        expect(parseStrToElements({ a: 'abc' })).toBe(null);
        expect(parseStrToElements(undefined)).toBe(null);
        expect(parseStrToElements(false)).toBe(null);
        expect(parseStrToElements(null)).toBe(null);
    });
    it('不含 HTML tag 的字串輸入, expect ["文字"]', () => {
        expect(parseStrToElements('testStr')).toEqual(
            expect.arrayContaining(['testStr'])
        );
        expect(parseStrToElements('中文字')).toEqual(
            expect.arrayContaining(['中文字'])
        );
    });
    it('包含 HTML tag 的字串輸入, expect 對應 element', () => {
        expect(parseStrToElements('<p>Test</p>')).toMatchSnapshot();
        expect(parseStrToElements('<div className="c1"><p>Test</p></div>')).toMatchSnapshot();
        expect(parseStrToElements(`<div className="c1"><h1>Test</h1>PureString~~~~</div>`)).toMatchSnapshot();
    });
    it('包含 空白、tab、換行 的字串輸入, expect 跳脫並顯示',  () => {
        expect(parseStrToElements('FirstLine\r\nSecondLine\r\nThirdLine')).toMatchSnapshot();
        expect(parseStrToElements('FirstLine\r\n   SecondLine')).toMatchSnapshot();
        expect(parseStrToElements('FirstLine\r\n\tSecondLine\r\n\tThirdLine')).toMatchSnapshot();
        expect(parseStrToElements(`<p>FirstLine\r\nSecondLine\r\n\tThirdLine\r\nFourthLine</p>`)).toMatchSnapshot();
    });
    it('包含可執行 inline js , expect 被移除',  () => {
        expect(parseStrToElements(`<button onClick="alert('hehehe')">Click Me</button>`)).toMatchSnapshot();
    });
    it('輸入 script 標籤與可執行 js, expect 被移除',  () => {
        expect(parseStrToElements(`<script>alert('hehehe');</script>`)).toMatchSnapshot();
    });
});