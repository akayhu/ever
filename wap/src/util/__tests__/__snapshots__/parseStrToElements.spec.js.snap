// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseStrToElements 包含 HTML tag 的字串輸入, expect 對應 element 1`] = `
Array [
  <p>
    Test
</p>,
]
`;

exports[`parseStrToElements 包含 HTML tag 的字串輸入, expect 對應 element 2`] = `
Array [
  <div>
    <p>
        Test
    </p>
</div>,
]
`;

exports[`parseStrToElements 包含 HTML tag 的字串輸入, expect 對應 element 3`] = `
Array [
  <div>
    <h1>
        Test
    </h1>
    PureString~~~~
</div>,
]
`;

exports[`parseStrToElements 包含 空白、tab、換行 的字串輸入, expect 跳脫並顯示 1`] = `
Array [
  "FirstLine",
  <br />,
  "
SecondLine",
  <br />,
  "
ThirdLine",
]
`;

exports[`parseStrToElements 包含 空白、tab、換行 的字串輸入, expect 跳脫並顯示 2`] = `
Array [
  "FirstLine",
  <br />,
  "
   SecondLine",
]
`;

exports[`parseStrToElements 包含 空白、tab、換行 的字串輸入, expect 跳脫並顯示 3`] = `
Array [
  "FirstLine",
  <br />,
  "
	SecondLine",
  <br />,
  "
	ThirdLine",
]
`;

exports[`parseStrToElements 包含 空白、tab、換行 的字串輸入, expect 跳脫並顯示 4`] = `
Array [
  <p>
    FirstLine
    <br />
    
    SecondLine
    <br />
    
    	ThirdLine
    <br />
    
    FourthLine
</p>,
]
`;

exports[`parseStrToElements 包含可執行 inline js , expect 被移除 1`] = `
Array [
  <button>
    Click Me
</button>,
]
`;

exports[`parseStrToElements 輸入 script 標籤與可執行 js, expect 被移除 1`] = `Array []`;
